src/Features/Core/Portable/UnusedReferences/UnusedReferencesService.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.UnusedReferences
+{
+    [ExportWorkspaceService(typeof(IUnusedReferencesService), ServiceLayer.Default), Shared]
+    internal class UnusedReferencesService : IUnusedReferencesService
+    {
+        private static readonly ReferenceType[] _processingOrder = new[]
+        {
+            ReferenceType.Project,
+            ReferenceType.Package,
+            ReferenceType.Assembly
+        };
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public UnusedReferencesService()
+        {
+
+        }
+
+        public async Task<ImmutableArray<ReferenceInfo>> GetUnusedReferencesAsync(
+            Project project,
+            ImmutableArray<ReferenceInfo> references,
+            CancellationToken cancellationToken)
+        {
+            // Create a lookup of used assembly paths
+            var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+            if (compilation is null)
+            {
+                return ImmutableArray<ReferenceInfo>.Empty;
+            }
+
+            var usedCompilationAssemblies = compilation.GetUsedAssemblyReferences(cancellationToken)
+                .Select(reference => reference.Display)
+                .OfType<string>();
+
+            return GetUnusedReferences(usedCompilationAssemblies, references);
+        }
+
+        internal static ImmutableArray<ReferenceInfo> GetUnusedReferences(
+            IEnumerable<string> usedCompilationAssemblies,
+            ImmutableArray<ReferenceInfo> references)
+        {
+            var usedAssemblyLookup = usedCompilationAssemblies.ToImmutableHashSet();
+
+            var unusedReferences = ImmutableArray.CreateBuilder<ReferenceInfo>();
+            var referencesByType = references.GroupBy(reference => reference.ReferenceType)
+                .ToImmutableDictionary(group => group.Key);
+
+            // We process the references in order by their type. This means we favor transitive references
+            // over direct references where possible.
+            foreach (var referenceType in _processingOrder)
+            {
+                if (!referencesByType.ContainsKey(referenceType))
+                {
+                    continue;
+                }
+
+                usedCompilationAssemblies = DetermineUnusedReferences(
```

Comment:
2 things: 

1. It seems weird that a method named `DeterminUnusedReferences` is being assigned to a variable named `usedCompilationAssemblies`
2. Can we make `usedCompilationAssemblies` a concrete type instead of `IEnumerable`? I know it's a nit but it could leave some weird semantics open if someone uses a different collection or streamed information of some kind.

Reply:
1. Update name to AddUnusedReferences and added a local `remainingUsedAssemblyLookup` to capture the return value which hopefully is more clear. #2 was addressed earlier.